feat(api): add `nvim_cmd`

famiu opened the pull request:
Adds the API function `nvim_cmd` which allows executing an Ex-command through a Dictionary which can have the same values as the return value of `nvim_parse_cmd()`. This makes it much easier to do things like passing arguments with a space to commands that otherwise may not allow it, or to make commands interpret certain characters literally when they otherwise would not.

Closes #14966

bfredl committed:
feat(keyset): allow hash conflicts

justinmk commented on src/nvim/api/vimscript.c:
```diff
@@ -736,3 +738,430 @@ Dictionary nvim_parse_expression(String expr, String flags, Boolean highlight, E
+    capture_ga = &capture_local;
+  }
+
+  try_start();
+  if (output) {
+    msg_silent++;
+  }
+
```
is TRY_WRAP useful here?

zeertzjq commented:
You can now use `string_as_cstr()` instead of `vim_strnsave()`.

famiu committed:
feat(api): add `nvim_cmd`

Adds the API function `nvim_cmd` which allows executing an Ex-command through a Dictionary which can have the same values as the return value of `nvim_parse_cmd()`. This makes it much easier to do things like passing arguments with a space to commands that otherwise may not allow it, or to make commands interpret certain characters literally when they otherwise would not.

clason commented:
Suggestion for follow-up PR: make `vim.cmd` call `nvim_cmd` for table arguments.

andmis commented on runtime/doc/api.txt:
```diff
@@ -1711,6 +1711,23 @@ nvim_call_function({fn}, {args})                        *nvim_call_function()*
                     Result of the function call
 
+nvim_cmd({*cmd}, {*opts})                                         *nvim_cmd()*
+                Executes an Ex-command.
```
Nit: there shouldn't be a dash between "Ex" and "command".

andmis commented on runtime/doc/api.txt:
```diff
@@ -1711,6 +1711,23 @@ nvim_call_function({fn}, {args})                        *nvim_call_function()*
+
+                Return: ~
+                    Command output (non-error, non-shell |:!|) if `output` is
+                    true, else empty string.
```
Any particular reason to return an empty string instead of `nil` in the latter case?

famiu replied:
We can only return a specific type from a function, C isn't dynamically typed.

andmis commented on runtime/doc/api.txt:
```diff
@@ -1711,6 +1711,23 @@ nvim_call_function({fn}, {args})                        *nvim_call_function()*
                 Return: ~
                     Result of the function call
 
+nvim_cmd({*cmd}, {*opts})                                         *nvim_cmd()*
```
It's confusing to have two functions in the API called `nvim_cmd` and `nvim_command` that do very similar things. Should we at least say in the docs a little bit about why there are two functions for this and summarize the difference between them or when you should prefer one over the other?

clason replied:
That's the way the API goes: functions that are added in a release cannot be removed (before 2.0). So some duplication is inevitable (and has happened before).

andmis replied:
That all sounds good, but then should we say a bit in the docs about why there are two functions and when people should prefer to use which?

andmis replied:
Resolved, TYVM!

andmis commented on runtime/doc/api.txt:
```diff
@@ -1711,6 +1711,23 @@ nvim_call_function({fn}, {args})                        *nvim_call_function()*
+                Return: ~
+                    Command output (non-error, non-shell |:!|) if `output` is
+                    true, else empty string.
+
```
How about a "See also:" pointing to `nvim_command`, and maybe adding `nvim_cmd` to the "See also:" of `nvim_command`?

andmis commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -175,6 +175,8 @@ enum {
 struct exarg {
   char *arg;                    ///< argument of the command
+  char **args;                  ///< position of command arguments
+  size_t argc;                     ///< number of command arguments
```
Nit: alignment.

andmis commented on src/nvim/ex_docmd.c:
```diff
@@ -1519,6 +1576,169 @@ bool parse_cmdline(char *cmdline, exarg_T *eap, CmdParseInfo *cmdinfo, char **er
   return true;
 }
 
+// Get end of an Ex-command argument with zero-indexed number, discarding trailing whitespace.
```
Nit: no dash in "Ex-command".

andmis replied:
<img width="262" alt="image" src="https://user-images.githubusercontent.com/5308024/167308826-b1d43089-b195-4f1c-87ed-c7873dfd400c.png">

andmis commented on src/nvim/ex_docmd.c:
```diff
@@ -1519,6 +1576,169 @@ bool parse_cmdline(char *cmdline, exarg_T *eap, CmdParseInfo *cmdinfo, char **er
+  return p;
+}
+
+/// Execute an Ex-command using parsed command line information.
```
Ditto.

bfredl commented on src/nvim/api/vimscript.c:
```diff
@@ -978,3 +988,377 @@ Dictionary nvim_parse_cmd(String str, Dictionary opts, Error *err)
+  OBJ_TO_BOOL(output, opts->output, false, "'output'");
+
+  // First, parse the command name and check if it exists and is valid.
+  if (!HAS_KEY(cmd->cmd)) {
+    VALIDATION_ERROR("No 'cmd' specified");
+  } else if (cmd->cmd.type != kObjectTypeString) {
+    VALIDATION_ERROR("'cmd' must be a String");
+  } else if (cmd->cmd.data.string.data[0] == NUL) {
+    VALIDATION_ERROR("'cmd' cannot be an empty String");
```
we don't need to be this detailed in validation errors. this can just be a single if with `||` and error message `'cmd' must be a non-empty string` (similar to existing errors like `x must be a non-negative integer`)

bfredl commented on src/nvim/api/vimscript.c:
```diff
@@ -978,3 +988,377 @@ Dictionary nvim_parse_cmd(String str, Dictionary opts, Error *err)
+    VALIDATION_ERROR("'cmd' must be a String");
+  } else if (cmd->cmd.data.string.data[0] == NUL) {
+    VALIDATION_ERROR("'cmd' cannot be an empty String");
+  } else {
```
no need for an else here, we return early on error, so after this should just be function indentation.

bfredl commented:
as a follow up: we should extend `kvec_t(char)` specifically with functions/macros like ga_concat and something like `kv_snprintf` (re-allocates buffer to fit formatted string), then we can get rid of IOSIZE and APPEND_MODIFIER and other such specific string manipulation macros.
